haskell/private/actions/link.bzl

```diff
@@ -36,6 +36,74 @@ def merge_parameter_files(hs, file1, file2):
     )
     return params_file
 
+def darwin_flags_for_linking_indirect_cc_deps(hs, cc, basename, dynamic):
+    """Write flags to force linking cc dependencies on MacOS to a parameter file.
+
+    GHC uses -dead_strip_dylibs in MacOS which discards all of the
+    indirect dependencies. Since cc libraries are underlinked, we use
+    -u symbol flags to force the linker to keep the cc dependencies
+    even when -dead_strip_dylibs is used.
+
+    Args:
+      hs: Haskell context.
+      cc: CcInteropInfo, information about C dependencies.
+      basename: The basename to use for the output file.
+      dynamic: Bool: Whether to link dynamically or statically.
+
+    Returns:
+      File: Parameter file with additional linker flags to be passed to GHC,
+      or None if we aren't building for darwin.
+    """
+
+    if not hs.toolchain.is_darwin:
+        return None
+
+    # On Darwin GHC will pass the dead_strip_dylibs flag to the linker. This
+    # flag will remove any shared library loads from the binary's header that
+    # are not directly resolving undefined symbols in the binary. I.e. any
+    # indirect shared library dependencies will be removed. This conflicts with
+    # Bazel's builtin cc rules, which assume that the final binary will load
+    # all transitive shared library dependencies. In particlar shared libraries
+    # produced by Bazel's cc rules never load shared libraries themselves. This
+    # causes missing symbols at runtime on MacOS, see #170.
+    #
+    # The following work-around applies the `-u` flag to the linker for an external
+    # symbol of each cc dependency. This forces the linker to resolve these
+    # undefined symbols in all the transitive shared cc library dependencies
+    # and keep the corresponding load commands in the binary's header.
+    suffix = ".dynamic.linker_flags" if dynamic else ".static.linker_flags"
+    linker_flags_file = hs.actions.declare_file(basename + suffix)
+
+    (_cc_static_libs, cc_dynamic_libs) = get_library_files(
+        hs,
+        cc.cc_libraries_info,
+        get_cc_libraries(cc.cc_libraries_info, cc.transitive_libraries),
+        dynamic,
+    )
+
+    # Some flags used in the invocation to nm are MacOS-specific and
+    # tested to exist at least since 10.15.
+    #
+    # -g     Display only global (external) symbols.
+    # -j     Just display the symbol names (no value or type).
+    # -p     Don't sort; display in symbol-table order.
+    # -U     Don't display undefined symbols.
+    hs.actions.run_shell(
+        inputs = cc_dynamic_libs,
+        outputs = [linker_flags_file],
+        command = """
+        touch {out}
+        for lib in {solibs}; do
+            {nm} -Ujpg "$lib" | head -1 | sed 's/^/-u /' >> {out}
+        done
+        """.format(
+            nm = cc.tools.nm,
+            solibs = " ".join(["\"" + l.path + "\"" for l in cc_dynamic_libs]),
+            out = linker_flags_file.path,
+        ),
+    )
```

Comment:
This is going to rerun whenever the PATH variable changes. Right?
And I guess the user needs to have `nm`, `head` and `sed` in the PATH.

Should we do anything about it?

Reply:
`nm` is provided by the toolchain.
You're correct on `head` and `sed`. You can pull those tools in from rules_sh's POSIX toolchain to avoid that. E.g. 
```head = posix.commands["head"]```

Reply:
> This is going to rerun whenever the PATH variable changes. Right?

No, Bazel controls `PATH` for this action.

Reply:
I've changed this to use the posix toolchain.